Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -17,6 +17,105 @@ import NIO
+public struct HTTPListHeaderIterator<T: Collection>: IteratorProtocol where T.Element == UInt8 {
+    
+    public typealias Element = ByteBufferView
+    
+    private var currentHeaderIndex: Int = -1
+    private var singleValueViewIterator: Array<ByteBufferView>.Iterator?
+    private let headerName: T
+    private let headers: HTTPHeaders
+    
+    fileprivate let comma = UInt8(",".utf8CString[0])
+    
+    /// Returns next index in headers
+    ///
+    /// - Parameter current: The index to begin iteration at
+    /// - Returns: The next index of the header in header array, or `nil` if not found
+    internal func headerIndex(after current: Int) -> Int? {
+        for (idx, currentHeader) in headers.headers.enumerated().dropFirst(current + 1) {
+            if headers.buffer.viewBytes(at: currentHeader.name.start,
+                                        length: currentHeader.name.length)
+                .compareReadableBytes(to: headerName) {
+                return idx
+            }
+        }
+        return nil
+    }
+
+    mutating public func next() -> ByteBufferView? {
+        if let next = self.singleValueViewIterator?.next() {
+            return next.trimSpaces()
+        } else {
+            // End of this buffer. Let's try to grab the next one.
+            guard let index = self.headerIndex(after: currentHeaderIndex) else {
+                // No more buffers left.
+                return nil
+            }
+            self.currentHeaderIndex = index
+            self.singleValueViewIterator = headers.buffer
+                .viewBytes(at: headers.headers[currentHeaderIndex].value.start,
+                           length: headers.headers[currentHeaderIndex].value.length)
+                .split(separator: comma)
+                .makeIterator()
+            return self.next()
+        }
+        
+    }
+    
+    init(headerName: T,
+         headers: HTTPHeaders) {
+        
+        self.headers = headers
+        self.headerName = headerName
+    }
+
+}
+
+extension HTTPHeaders {
+    
+    private static let connectionString = "connection".utf8
+    
+    private static let keepAliveString = "keep-alive".utf8
+    private static let closeString = "close".utf8
+    
+    internal var isKeepAlive: Bool? {
+        get {
+            var keepAlive = false
+            var close = false
+            
+            var tokenizer = HTTPListHeaderIterator(
+                headerName: HTTPHeaders.connectionString, headers: self)
+            
+            while let nextToken = tokenizer.next() {
+                keepAlive = keepAlive ||
+                    (nextToken.compareReadableBytes(to: HTTPHeaders.keepAliveString))
+                close = close ||
+                    (nextToken.compareReadableBytes(to: HTTPHeaders.closeString))
+            }
+            
+            // TODO: Handle the case where both keep-alive and close are used
+            if keepAlive { return true }
+            if close { return false }
```

Comment:
Let's clear this up a bit. I agree with your decision not to do error handling yet, but if we're not going to do error handling then we don't need to read the whole header block. Instead, we can exit early once we see the first token:

```swift
while let nextToken = tokenizer.next() {
    if nextToken.compareReadableBytes(to: HTTPHeaders.keepAliveString) { return .keepAlive }
    else if nextToken.compareReadableBytes(to: HTTPHeaders.closeString) { return .close }
}

return .unspecified
```